ext/rotoscope/callsite.c

```diff
@@ -1,28 +1,43 @@
 #include "callsite.h"
 #include <ruby.h>
 #include <ruby/debug.h>
+#include <stdbool.h>
+
+static VALUE caller_frame(int *line, bool ruby_call) {
+  VALUE frames[2] = {Qnil, Qnil};
+  int lines[2] = {0, 0};
+
+  int frame_index = ruby_call ? 1 : 0;
+
+  // There is currently a bug in rb_profile_frames that
+  // causes the start argument to effectively always
+  // act as if it were 0, so we need to also get the top
+  // frame. (https://bugs.ruby-lang.org/issues/14607)
+  rb_profile_frames(0, frame_index + 1, frames, lines);
+
+  *line = lines[frame_index];
+  return frames[frame_index];
+}
 
 rs_callsite_t c_callsite(rb_trace_arg_t *trace_arg) {
-  VALUE path = rb_tracearg_path(trace_arg);
+  int line;
+  VALUE frame = caller_frame(&line, false);
   return (rs_callsite_t){
-      .filepath = path, .lineno = FIX2INT(rb_tracearg_lineno(trace_arg)),
+      .filepath = rb_tracearg_path(trace_arg),
+      .lineno = FIX2INT(rb_tracearg_lineno(trace_arg)),
+      .method_name = rb_profile_frame_method_name(frame),
+      .singleton_p = rb_profile_frame_singleton_method_p(frame),
   };
 }
 
 rs_callsite_t ruby_callsite() {
```

Comment:
Can you add a minimal comment explaining the difference between `ruby_callsite` and `c_callsite`?

Reply:
I would like to get rid of them as two separate code paths, but I think the `bool ruby_call` parameter still deserves documenting.

The reason they diverged historically was because `rb_tracearg_` API for the callsite was working for C but not ruby. That tracepoint API was trying to returning the code location just after the call (i.e. inside the called function) but it would also skip non-ruby stack frames.  `rb_profile_frames` replaced some earlier dirty hack we had isolated to `ruby_callsite` and allows us to get more than just the current code location, allowing us to use index `0` to continue getting the correct location for C but `1` to skip a frame and get the caller location for ruby.

Reply:
Added a comment to `caller_frame` on this key difference.